core/subscription/subscription_engine.hpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_SUBSCRIPTION_ENGINE_HPP
+#define KAGOME_SUBSCRIPTION_ENGINE_HPP
+
+#include <list>
+#include <memory>
+#include <shared_mutex>
+#include <unordered_map>
+
+namespace kagome::subscription {
+  template <typename Key, typename Type, typename... Arguments>
+  class Subscriber;
+
+  template <typename Key, typename Type, typename... Arguments>
+  class SubscriptionEngine final
+      : public std::enable_shared_from_this<
+            SubscriptionEngine<Key, Type, Arguments...>> {
+   public:
+    using KeyType = Key;
+    using ValueType = Type;
+    using SubscriberType = Subscriber<KeyType, ValueType, Arguments...>;
+    using SubscriberPtr = std::shared_ptr<SubscriberType>;
+    using SubscriberWPtr = std::weak_ptr<SubscriberType>;
+
+    /// List preferable here because of remains the whole containers iterators
+    /// alive after remove from the middle
+    /// TODO(iceseer): remove processor cache penalty, while iterating, using
+    /// custom allocator
+    using SubscribersContainer = std::list<SubscriberWPtr>;
+    using IteratorType = typename SubscribersContainer::iterator;
+
+   private:
+    template <typename KeyType, typename ValueType, typename... Args>
+    friend class Subscriber;
+    using KeyValueContainer = std::unordered_map<KeyType, SubscribersContainer>;
+
+    std::shared_mutex subscribers_map_cs_;
```

Comment:
For what reason is `shared_mutex` used?

Reply:
This mutex guards ```subscribers_map_```. The changes in this resource are made only in ```subscribe``` and ```unsubscribe``` methods. ```notify``` doesn't need write-lock, it only reads and can do this in multiple-threads in theory.

Reply:
But I didn't find using special feature of `shared_mutex`, e.g. `lock_shared`. Without it using of `shared_mutex` useless than `mutex`

Reply:
```    void notify(KeyType const &key, Arguments const &... args) {
      std::shared_lock lock(subscribers_map_cs_);
      auto it = subscribers_map_.find(key);
      if (subscribers_map_.end() == it) return;```